renderdoc/driver/vulkan/wrappers/vk_resource_funcs.cpp

```diff
@@ -880,12 +883,13 @@ VkResult WrappedVulkan::vkFlushMappedMemoryRanges(VkDevice device, uint32_t memR
         if(size == VK_WHOLE_SIZE)
           size = state->mapOffset + state->mapSize - offs;
 
-        GetResourceManager()->MarkMemoryFrameReferenced(GetResID(pMemRanges[i].memory), offs, size,
-                                                        eFrameRef_CompleteWrite);
+        GetResourceManager()->MarkMemoryFrameReferenced(memid, offs, size, eFrameRef_CompleteWrite);
```

Comment:
The issue here is that `MarkMemoryFrameReferenced` is not decaying the reference type when it passes it along to `MarkResourceFrameReferenced` so the whole memory object will wrongly be marked as `CompleteWrite` even if only some of the memory is being updated. The same is true in a couple of other places that use `CompleteWrite` on regions of buffers.

The `CompleteWrite` here is implicitly scoped by the offset and size, which means that _only in that region_ does it completely write. We still want to preserve that per-interval information since that allows the replay side resetting do a fast-clear on those regions, but when `MarkMemoryFrameReferenced` forwards to `MarkResourceFrameReferenced` for the whole resource, it will have to decay a `CompleteWrite` to a `PartialWrite` when only some of the region is modified. Since it doesn't have easy access to the size of the whole memory object, we'll need to pass that in I think.

Tagging @bjoeris to take a look at this and give any further input.

Reply:
This issue is still present I think. Taking a concrete example:

A memory is 4MB in size, and the first thing in the frame is that the region at offset 1MB and size 1MB is mapped and flushed. This marks that region as `CompleteWrite`, but currently `MarkMemoryFrameReferenced` will then mark the whole resource as referenced by `CompleteWrite` as a result because there was no previous reference. Instead it should be decaying the reference for the whole resource to `PartialWrite` when the region doesn't cover the whole resource, since a `CompleteWrite` to a sub-segment is not enough for us to consider skipping that resource.

Reply:
Ah, I had misunderstood the original comment when I replied. I think this is actually a bug in `MarkMemoryFrameReferenced`: currently, if only a portion of a memory resource is accessed as `CompleteWrite`, then the `m_FrameReferencedResources` will incorrectly be updated to `CompleteWrite`. However, I don't think this would have previously caused any visible symptoms, because, if any portion of the memory was later read, that read would replace `CompleteWrite` in `m_FrameReferencedResources` (because the old and new reference types are composed using `ComposeFrameRefsDisjoint`).

When I added the subresource tracking (memory ranges and image subresources), I tried to change most of the code that was using the per-resource frame refs to use the per-subresource frame refs for images and memory. For images and memory, the calls to `MarkResourceFrameReferenced` are just intended to update `m_FrameReferencedResources` to maintain a conservative upper bound on the overall access type for the entire resource.

Reply:
Right that's what I was meaning - arguably a bug before but only significant now because we're attaching more semantic information to that state/reference. I'm not sure if you could construct a scenario where this would completely break replay even with these changes but it does feel like it should be fixed to avoid confusion or potential problems later.

Reply:
I have put in a hack fix to lower a CompleteWrite down to PartialWrite if the offset and size doesn't cover the entire memory. I'm wondering maybe we should decouple MarkMemoryReferenced with MarkResourceFrameReferenced?

Reply:
I think that fix looks fine to me, decoupling those functions I don't think would be a good idea since as Benson mentioned above that would just mean that every call site has to call both functions. They are very much the same thing, and `MarkMemoryReferenced` can be considered a specialisation of `MarkResourceFrameReferenced`, it just has to properly decay the reference in this way.

Reply:
That change in `MarkMemoryFrameReferenced` looks good to me.